Translating reference-react-component (#130)
reactjs/pt-BR.reactjs.org: content/docs/reference-react-component.md

  - "tips/use-react-with-other-libraries.html"
---

> Esta página contem uma referência detalhada da API para a definição de classes de componentes React. Nós assumimos que você possui familiaridade com conceitos fundamentais do React, como [Componentes e Props](/docs/components-and-props.html), bem como [State e Lifecycle](/docs/state-and-lifecycle.html). Se isto não é familiar para você, leia essas páginas primeiro.

Reviewer: `Lifecycle` acho q seria traduzido, porque na pagina dele esta `State e Ciclo de Vida`!

```suggestion
Esta página contém uma referência detalhada da API para a definição de classes de componentes React. Nós assumimos que você possui familiaridade com conceitos fundamentais do React, como [Componentes e Props](/docs/components-and-props.html), bem como [State e Ciclo de vida](/docs/state-and-lifecycle.html). Se isto não é familiar para você, leia essas páginas primeiro.
```

  }
}
```

O único método que você *deve* definir em uma subclasse de `React.Component` é chamado [`render()`](#render). Todos os outros métodos descritos nesta página são opcionais.

**Nós somos fortemente contra a criação de seus próprios componentes base.** Em componentes React, [O reuso do código é obtido primariamente através de composição ao invés de herança]
(/docs/composition-vs-inheritance.html).

>Nota:
>
>React não te obriga a utilizar a sintaxe ES6 para classes. Se preferir não usá-la, você pode usar o módulo `create-react-class` ou alguma outra abstração similar. Dê uma olhada em 
[Usando React sem ES6](/docs/react-without-es6.html) para mais sobre este assunto.

### O Ciclo de vida de um Componente {#component-life-cycle}

Cada componente possui muitos "métodos do ciclo de vida" que você pode sobrescrever para executar determinado código em momentos particulares do processo. **Você pode usar [este diagrama do ciclo de vida](http://projects.wojtekmaj.pl/react-lifecycle-methods-diagram/) para consulta.** Na lista abaixo, os métodos do ciclo de vida mais usados estão marcados em **negrito**. O resto deles, existe para o uso em casos relativamente raros.

> #### Montando {#mounting}

Reviewer 2: @cezaraugusto vamos traduzir a palavra `Mounting`? Tou com dúvida aqui

cc: @halian-vilela @tibuurcio

Author: Aguardando a resposta de vcs @cezaraugusto , @halian-vilela e @tibuurcio

Reviewer 3: @glaucia86 @rcantuario 
Eu acho que seria o caso de deixar o original entre parênteses ,só para facilitar que o leitor faça a ligação com os nomes dos métodos de lifecycle.

Author: Beleza, alterarei

> #### Montando {#mounting}

Reviewer 3: ```suggestion
#### Montando (mounting) {#mounting}
```

- [**`constructor()`**](#constructor)
- [`static getDerivedStateFromProps()`](#static-getderivedstatefromprops)
- [**`render()`**](#render)
- [**`componentDidMount()`**](#componentdidmount)

>Nota:
>
>Estes métodos são considerados legado e você deve [evitá-los]/blog/2018/03/27/update-on-async-rendering.html) em código novo:
>
>- [`UNSAFE_componentWillMount()`](#unsafe_componentwillmount)

#### Atualizando {#updating}

Uma atualização pode ser causada por alterações em props ou no state. Estes métodos são chamados na seguinte ordem quando um componente esta sendo re-renderizado:

- [`static getDerivedStateFromProps()`](#static-getderivedstatefromprops)
- [`shouldComponentUpdate()`](#shouldcomponentupdate)
- [**`render()`**](#render)
- [`getSnapshotBeforeUpdate()`](#getsnapshotbeforeupdate)
- [**`componentDidUpdate()`**](#componentdidupdate)

>Nota:
>
>Estes métodos são considerados legado e você deve [evitá-los]/blog/2018/03/27/update-on-async-rendering.html) em código novo:
>
>- [`UNSAFE_componentWillUpdate()`](#unsafe_componentwillupdate)
>- [`UNSAFE_componentWillReceiveProps()`](#unsafe_componentwillreceiveprops)

> #### Desmontando {#unmounting}

Reviewer 2: @cezaraugusto mesma coisa q o caso do `Mouting`. Traduziremos essa palavra também? `Unmounting`?

cc: @tibuurcio @cezaraugusto @halian-vilela

Author: Aguardando também =)

Reviewer 3: Idem acima. Por mim, a gente colocava o original entre parênteses

Author: Vou alterar.. Valeu

> #### Desmontando {#unmounting}

Reviewer 3: ```suggestion
#### Desmontando (unmounting) {#unmounting}
```

Estes métodos são chamados quando um componente está sendo removido do DOM:

- [**`componentWillUnmount()`**](#componentwillunmount)

> #### Tratando erros {#handling errors}

Reviewer 2: ```suggestion
#### Tratando erros {#error-handling}
```

não alterar os Id's @rcantuario

Estes métodos são chamados quando existir um erro durante a renderização, em um método do ciclo de vida, ou no construtor de qualquer componente-filho.

- [`static getDerivedStateFromError()`](#static-getderivedstatefromerror)
- [`componentDidCatch()`](#componentdidcatch)

### Outras APIs {#other-apis}

Cada componente também fornece outras APIs:

  - [`setState()`](#setstate)
  - [`forceUpdate()`](#forceupdate)

### Propriedades de Classes {#class-properties}

  - [`defaultProps`](#defaultprops)
  - [`displayName`](#displayname)

### Propriedades da instância {#instance-properties}

  - [`props`](#props)
  - [`state`](#state)

* * *

> ## Referencia {#reference}

Reviewer 3: ```suggestion
## Referência {#reference}
```

> ### Métodos mais usados do Ciclo de Vida {#métodos-mais-usados-do-ciclo-de-vida}

Reviewer: Não traduzimos as "âncoras" e adicionei um enter abaixo do titulo como no arquivo original.

```suggestion
### Métodos mais usados do Ciclo de Vida {#commonly-used-lifecycle-methods}

```

### `render()` {#render}

```javascript
render()
```
O método `render()` é o único método obrigatório em um class-component.

Quando chamado, ele examina `this.props` e `this.state` e retorna um dos seguintes tipos:

- **Elementos React.** Tipicamente criados via [JSX](/docs/introducing-jsx.html). Por exemplo, `<div />` e `<MyComponent />` são elementos React que instruem o React para renderizar um nó do DOM, ou outro componente definido pelo usuário, respectivamente.
- **Arrays e fragmentos.** Permitem que você retorne múltiplos elementos ao renderizar. Veja a documentação em [fragments](/docs/fragments.html) para mais detalhes.
- **Portals**. Permitem que você renderize componentes-filhos em uma sub-árvore diferente do DOM. Veja a documentação em [portals](/docs/portals.html) para mais detalhes.
- **String e números.** Estes são renderizados como nós de texto no DOM.
- **Booleanos ou `null`**. Não renderizam nada.(A maioria existe para suportar  o padrão `return test && <Child />` , onde `test` é um booleano.)

> A função `render()` deve ser pura, o que significa que ela não modifica o state, ela retorna o mesmo resultado a cada vez que é chamada, e isso não interage diretamente com o browser.

Reviewer 2: ```suggestion
A função `render()` deve ser pura, o que significa que ela não modifica o state. Pois, ela retorna o mesmo resultado a cada vez que é chamada e isso não interage diretamente com o browser.
```

Se você precisar interagir com o browser, faça isto no método `componentDidMount()` ou em outros métodos do ciclo de vida. Manter `render()` puro faz com que os componentes sejam fáceis de se trabalhar.

> Nota
>
> `render()` não será invocado se  [`shouldComponentUpdate()`](#shouldcomponentupdate) retornar false.

* * *

> ### `construtor()` {#constructor}

Reviewer: Acho que aqui como se refere ao método não deveria traduzir, oq acha?

```suggestion
### `constructor()` {#constructor}
```

```javascript
constructor(props)
```

**Se você não inicializar o state e não fizer o bind dos métodos, você não precisa implementar um construtor para seu componente**

O construtor para um componente React é chamado antes que este componente seja montado. Quando estiver implementando o construtor para uma subclasse de `React.Component`, você deveria chamar `super(props)` antes de qualquer outra coisa.
Caso contrário `this.props` será indefinido no construtor, o que pode levar a bugs.

> Tipicamente, em React, os construtores são usados somente para dois propósitos:

Reviewer 2: ```suggestion
Normalmente, em React, os construtores são usados somente para dois propósitos:
```
Vamos usar o sinônimo de `tipicamente`? => normalmente? Acho q vai ficar mais lindo :)

https://www.sinonimos.com.br/tipicamente/

cc: @rcantuario

Author: De acordo =)

* Inicializar [local state](/docs/state-and-lifecycle.html) através da atribuição de um objeto a `this.state`.
* Ligação (binding) de um método [manipulador de eventos](/docs/handling-events.html) à uma instância.

Você **não deve chamar `setState()`**  no `constructor()`. Ao invés disso, se seu componente precisa de local state, **atribua o initial state à `this.state`** diretamente no construtor:

```js
constructor(props) {
  super(props);
  // Não chame this.setState() aqui!
  this.state = { counter: 0 };
  this.handleClick = this.handleClick.bind(this);
}
```

> O método consturtor é o único lugar onde você deve atribuir `this.state` diretamente. Em todos os outros métodos, você precisa usar `this.setState()`.

Reviewer 2: ```suggestion
O método construtor é o único lugar onde você deve atribuir `this.state` diretamente. Em todos os outros métodos, você precisa usar `this.setState()`.
```

Evite introduzir qualquer efeito colateral no construtor. Para estes casos use `componentDidMount()`.

>Nota
>
>**Evite copiar props no state! Este é um erro comum:**
>
>```js
>constructor(props) {
>  super(props);
>  // Não faça isso!
>  this.state = { color: props.color };
>}
>```
>

> >O problema é que isto é desnecessário (você pode usar `this.props.color` diretamente), e cria bugs (atualizações em `color` não serão refletidas no state).

Reviewer 2: ```suggestion
O problema é que isto é desnecessário (você pode usar `this.props.color` diretamente) e cria bugs (atualizações em `color` não serão refletidas no state).
```

Reviewer: @glaucia86 acho q a "seta (`>`)" no  inicio da frase permanece, pois é uma `Nota`

componentDidMount()
```

`componentDidMount()` É invocado imediatamente após um componente ser montado (inserido na árvore). Inicializações que exijam nós do DOM devem vir aqui. Se precisar carregar data de um endpoint remoto, este é um bom lugar para instanciar sua requisição.

Este método é um bom lugar para colocar qualquer subscrição. Se fizer isto, não esqueça de desinscrever no `componentWillUnmount()`.

> Você **pode chamar `setState()` diretamente** dentro do `componentDidMount()`. Ele irá disparar uma renderização extra, mas isto irá ocorrer antes que o browser atualize a tela. Isso garante que mesmo que o `render()` seja chamado duas vezes neste caso, o usuário não verá o state intermediário. Use este pattern com cuidado porque isto frequentemente causa issues de performance. Na maioria dos casos, você deve atribuir o initial state no `constructor()`. Isto pode, no entanto, ser necessário para casos como modais e tooltips quando você precisa mensurar um nó do DOM antes de renderizar algo que dependa de seu tamanho ou posição.

Reviewer 2: ```suggestion
Você **pode chamar `setState()` diretamente** dentro do `componentDidMount()`. Ele irá disparar uma renderização extra. Porém, isto ocorrerá antes que o browser atualize a tela. Isso garante que mesmo que o `render()` seja chamado duas vezes neste caso, o usuário não verá o state intermediário. Use este pattern com cuidado porque isto frequentemente causa issues de performance. Na maioria dos casos, você deve atribuir o initial state no `constructor()`. Isto pode, no entanto, ser necessário para casos como modais e tooltips quando você precisa mensurar um nó do DOM antes de renderizar algo que dependa de seu tamanho ou posição.

```

Reviewer: Acho que essa linha extra no final da sua sugestão pode ser removida @glaucia86

componentDidUpdate(prevProps, prevState, snapshot)
```

`componentDidUpdate()` é invocado imediatamente após alguma atualização ocorrer. Este método não é chamado pelo initial render.

Use isto como uma oportunidade para alterar o DOM quando o componente for atualizado. Este também é um bom lugar para realizar requisições de rede enquanto compara as props atuais com as props anteriores (e.g. uma chamada de rede pode não ser necessária se as props não mudaram).

```js
componentDidUpdate(prevProps) {
  // Uso típico, (não esqueça de comparar as props):
  if (this.props.userID !== prevProps.userID) {
    this.fetchData(this.props.userID);
  }
}
```

> Você **pode chamar `setState()` imediatamente** dentro do `componentDidUpdate()` mas perceba que **isto deve estar encapsulado em uma condição** como no exemplo abaixp, ou você irá causar um loop infinito. Isto também causaria uma re-renderização extra, que por mais que não seja visível para o usuário pode afetar a performance do componente. Se você está tentando "espelhar" algum state para uma prop vinda de cima, considere usar a prop diretamente. Leia mais sobre [porque copiar props no state causa bugs](/blog/2018/06/07/you-probably-dont-need-derived-state.html).

Reviewer 2: ```suggestion
Você **pode chamar `setState()` imediatamente** dentro do `componentDidUpdate()` mas perceba que **isto deve estar encapsulado em uma condição** como no exemplo abaixo ou você irá causar um loop infinito. Isto também causaria uma re-renderização extra, que por mais que não seja visível para o usuário pode afetar a performance do componente. Se você está tentando "espelhar" algum state para uma prop vinda de cima, considere usar a prop diretamente. Leia mais sobre [porque copiar props no state causa bugs](/blog/2018/06/07/you-probably-dont-need-derived-state.html).
```

Reviewer 3: @glaucia86 Eis aqui um exemplo daquele bug do GH q eu tinha falado! Veja que não tem highlight no que você corrigiu!

Reviewer 2: Eita!

> Se seu componente implementa o método `getSnapshotBeforeUpdate()` (o que é raro), o valor que ele retorna será passado como o terceiro parâmetro "snapshot" para `componentDidUpdate()`. Caso contrário este parâmetro será undefined.

Reviewer 2: ```suggestion
Se o seu componente implementa o método `getSnapshotBeforeUpdate()` (o que é raro), o valor que ele retorna será passado como o terceiro parâmetro "snapshot" para `componentDidUpdate()`. Caso contrário, este parâmetro será undefined.
```

componentWillUnmount()
```

> `componentWillUnmount()` é invocado imediatamente antes que um componente seja desmontado e destruído. Qualquer limpeza necessária deve ser realizada neste método, como invalidar timers, cancelar requisições de rede, ou limpar qualquer subscrição que foi criada no `componentDidMount()`.

Reviewer 2: ```suggestion
`componentWillUnmount()` é invocado imediatamente antes que um componente seja desmontado e destruído. Qualquer limpeza necessária deve ser realizada neste método, como invalidar timers, cancelar requisições de rede ou limpar qualquer subscrição que foi criada no `componentDidMount()`.
```

`shouldComponentUpdate()` é invoado antes do rendring quando novas props ou state são recebigos. O valor default é `true`. Este método não é chamado na renderização inicial ou quando `forceUpdate()`é usado .

Este método existe somente para **[otimização de performance ](/docs/optimizing-performance.html).** Não confie nele para "evitar" a renderização, pois isso pode levar a bugs. **Considere usar [`PureComponent`](/docs/react-api.html#reactpurecomponent)** no lugar de escrever `shouldComponentUpdate()` manualmente. `PureComponent` realiza uma comparação superficial das props e do state, e reduz as chances the pular um update necessário.

Se você está confiante que quer escrever isto manualmente, pode comparar `this.props` com `nextProps` e `this.state` com `nextState`
e retornar `false` para informar o React que o update pode ser pulado. Perceba que retornando `false` não evita que componentes filhos sejam renderizados novamente quando o state *deles* sofrer alterações.

Não recomendamos fazer verificações de igualdade profundas ou usar `JSON.stringify()` dentro de `shouldComponentUpdate()`. Isto é ineficiente e irá prejudicar a performance.

> Atualmente, se `shouldComponentUpdate()` retornar `false`, então [`UNSAFE_componentWillUpdate()`](#unsafe_componentwillupdate), [`render()`](#render), e [`componentDidUpdate()`](#componentdidupdate) não serão invocados. No futuro, React pode tratar  `shouldComponentUpdate()` como uma dica ao invés de uma rigorosa diretiva. e retornar `false` pode continuar resultando em re-renderização do componente.

Reviewer 2: ```suggestion
Atualmente, se `shouldComponentUpdate()` retornar `false`, então [`UNSAFE_componentWillUpdate()`](#unsafe_componentwillupdate), [`render()`](#render) e [`componentDidUpdate()`](#componentdidupdate) não serão invocados. No futuro, React pode tratar  `shouldComponentUpdate()` como uma dica ao invés de uma rigorosa diretiva e retornar `false` pode continuar resultando em re-renderização do componente.
```

static getDerivedStateFromProps(props, state)
```

> `getDerivedStateFromProps` é invocado imediatamente antes de chamar o método render, ambos na montagem inicial e nas atualizações subsequente. Devem retornar um objeto para atualizar o state, ou null para não atualizar nada.

Reviewer 2: ```suggestion
`getDerivedStateFromProps` é invocado imediatamente antes de chamar o método render, ambos na montagem inicial e nas atualizações subsequente. Devem retornar um objeto para atualizar o state ou null para não atualizar nada.
```

> Este método existe para [casos de uso raros] (/blog/2018/06/07/you-probably-dont-need-derived-state.html#when-to-use-derived-state) onde o state depende de mudanças nas props ao longo do tempo. Por exemplo, pode ser útil para implementar um componente `<Transition>` que compara seus filhos anteriores e próimos para decidir quais deles devem ser animados.

Reviewer 2: ```suggestion
Este método existe para [casos de uso raros] (/blog/2018/06/07/you-probably-dont-need-derived-state.html#when-to-use-derived-state) onde o state depende de mudanças nas props ao longo do tempo. Por exemplo, pode ser útil para implementar um componente `<Transition>` que compara seus filhos anteriores e próximos para decidir quais deles devem ser animados.
```

> Derivando o state leva a código verboso e faz seus componentes difíceis de compreender.

Reviewer 2: ```suggestion
Derivando o state deixa o código verboso e deixa seus componentes difíceis de compreender.
```

[Tenha certeza de estar familiarizado com alternativas mais simples:](/blog/2018/06/07/you-probably-dont-need-derived-state.html)

* Se precisar  **perform a side effect** (por exemplo, buscar dados ou uma animação) em resposta a uma alteração em props, use [`componentDidUpdate`](#componentdidupdate) no lugar.

* Se você quer **re-compute some data only when a prop changes**, [use um auxiliar de memorização no lugar](/blog/2018/06/07/you-probably-dont-need-derived-state.html#what-about-memoization).

> * ISe você quer **"resetar" o state quando uma prop mudar**, considere criar um componente [completamente controlado](/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-controlled-component) ou [completamente controlado com uma `chave`](/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key) instead.

Reviewer 2: ```suggestion
Se você quiser **"resetar" o state quando uma prop mudar**, considere criar um componente [completamente controlado](/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-controlled-component) ou [completamente controlado com uma `chave`](/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key) instead.
```

> Este método não potem acesso a instância do componente. Se você quiser, pode reusar o código entre o método `getDerivedStateFromProps()` e os métodos de outra classe extraindo funções puras para as props e state do componente, fora da definição da classe.

Reviewer 2: ```suggestion
Este método não possuem acesso a instância do componente. Se você quiser, pode reusar o código entre o método `getDerivedStateFromProps()` e os métodos de outra classe extraindo funções puras para as props e state do componente fora da definição da classe.
```

getSnapshotBeforeUpdate(prevProps, prevState)
```

`getSnapshotBeforeUpdate()` é invocado imediatamente antes que o retorno da renderização mais recente seja escrito e.g. no DOM. Isto permite que o componente capture alguma informação do DOM (e.g. posição do scroll) antes que ela seja potencialmente alterada. Qualquer valor retornado por este método do ciclo de vida será passado como parâmetro para `componentDidUpdate()`.

> Este caso de uso não é comum, mas pode ocorrer em UIs como um thread de um chat que precise tratar a posição do scroll de uma maneira especial.

Reviewer 2: ```suggestion
Este caso de uso não é comum. Porém, pode ocorrer em UIs como um thread de um chat que precise tratar a posição do scroll de uma maneira especial.
```

}
```

> Nota
>

> > `getDerivedStateFromError()` é chamado durante a fase de renderização, portanto efeitos colaterais (*side-effects*) não são permitidos.

Reviewer 2: ```suggestion
 `getDerivedStateFromError()` é chamado durante a fase de renderização. Portanto, os efeitos colaterais (*side-effects*) não são permitidos.
```

Reviewer: aqui também a "seta (`>`)" deve permanecer.

componentDidCatch(error, info)
```

Este lifecycle é invocado após um erro ter sido lançado por um componente descendente.
Ele recebe dois parâmetros:

1. `error` - O erro que foi lançado.
2. `info` - Um objeto com uma chave `componentStack` contendo [informações sobre qual componente lançou o erro](/docs/error-boundaries.html#component-stack-traces).

> `componentDidCatch()` é chamado durante a fase de "commit", portanto efeitos colaterais (*side-effects*) não são permitidos.

Reviewer 2: ```suggestion
`componentDidCatch()` é chamado durante a fase de "commit". Portanto, os efeitos colaterais (*side-effects*) não são permitidos.
```

}
```

> Nota
>

> > No evento de um erro, você pode renderizar uma UI de *fallback* com `componentDidCatch()` chamando `setState`, mas isto será depreciado numa *release* futura.

Reviewer 2: ```suggestion
 No evento de um erro, você pode renderizar uma UI de *fallback* com `componentDidCatch()` chamando `setState`. Mas, isto será depreciado numa *release* futura.
```

Reviewer: "seta (`>`)"

> * Se você usa `componentWillReceiveProps` para **recomputar algum dado somente quando uma *prop* muda**, [utilize um *memoization helper*](/blog/2018/06/07/you-probably-dont-need-derived-state.html#what-about-memoization).
> * Se você usa `componentWillReceiveProps` para **"resetar" algum *state* quando uma *prop* muda**, considere ou criar um componente [completamente controlado](/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-controlled-component) ou [completamente não controlado com uma `key`](/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key) como alternativa.
> Para outros casos de uso, [siga as recomendações neste *post* do *blog* sobre *derived state*](/blog/2018/06/07/you-probably-dont-need-derived-state.html).

`UNSAFE_componentWillReceiveProps()` é invocado antes que um componente montado receba novas *props*. Se você precisa atualizar o estado em resposta a mudanças na *prop* (por exemplo, para resetá-lo), você pode comparar `this.props` e `nextProps` e realizar transições de *state* utilizando `this.setState()` neste método.

> Note que se um componente pai causar a rerenderização do seu componente, este método será chamado mesmo se as *props* não foram alteradas. Certifique-se de comparar o valor atual e o próximo se você deseja somente manipular mudanças.

Reviewer 2: ```suggestion
Note que, se um componente pai causar a rerenderização do seu componente, este método será chamado mesmo se as *props* não foram alteradas. Certifique-se de comparar o valor atual e o próximo se você deseja somente manipular mudanças.
```

UNSAFE_componentWillUpdate(nextProps, nextState)
```

> Nota
>
> Este lifecycle era nomeado `componentWillUpdate`. Este nome continuará a funcionar até a versão 17. Utilize o [codemod `rename-unsafe-lifecycles`](https://github.com/reactjs/react-codemod#rename-unsafe-lifecycles) para atualizar automaticamente seus componentes.

`UNSAFE_componentWillUpdate()` é invocado antes da renderização quando novas *props* ou *state* estão sendo recebidos. Utilize este método como uma oportunidade para realizar preparações antes que uma atualização ocorra. Este método não é chamado para a renderização inicial.

Note que você não pode chamar `this.setState()` aqui; e nem deveria fazer nada além (por exemplo, realizar o *dispatch* de uma *action* do *Redux*) que desencadearia uma atualização em um componente React antes que `UNSAFE_componentWillUpdate()` retorne.

> Tipicamente, este método pode ser substituído por `componentDidUpdate()`. Se você estiver lendo do *DOM* neste método (por exemplo, para salvar a posição de rolagem), você pode mover esta lógica para `getSnapshotBeforeUpdate()`.

Reviewer 2: ```suggestion
Normalmente, este método pode ser substituído por `componentDidUpdate()`. Se você estiver lendo do *DOM* neste método (por exemplo, para salvar a posição de rolagem), você pode mover esta lógica para `getSnapshotBeforeUpdate()`.
```

> Nota
>
> `UNSAFE_componentWillUpdate()` não será invocado se [`shouldComponentUpdate()`](#shouldcomponentupdate) retornar *false*.

* * *

## Outras APIs {#other-apis-1}

Diferentemente dos métodos do *lifecycle* acima (que o React chama por você), os métodos abaixo são métodos que *você* pode chamar a partir de seus componentes.

Existem apenas dois deles: `setState()` e `forceUpdate()`.

### `setState()` {#setstate}

```javascript
setState(updater[, callback])
```

> `setState()` enfileira mudanças ao *state* do componente e diz ao React que este componente e seus componentes filho precisam ser rerenderizados com a atualização do *state*. Este é o principal método que você utiliza para atualizar a UI em resposta a *event handlers* e à resposta de servidores.

Reviewer 2: ```suggestion
`setState()` enfileira mudanças ao *state* do componente e diz ao React que este componente e seus componentes filhos precisam ser re-renderizados com a atualização do *state*. Este é o principal método que você utiliza para atualizar a UI em resposta a *event handlers* e à resposta de servidores.
```

> Pense em `setState()` como uma *requisição* ao invés de um comando imediato para atualizar o componente. Para uma melhoria na performance, o React pode atrasar a atualização, e então atualizar diversos componentes numa só leva. O React não garante que as mudanças no *state* são aplicadas imediatamente.

Reviewer 2: ```suggestion
Pense em `setState()` como uma *requisição* ao invés de um comando imediato para atualizar o componente. Para uma melhoria na performance, o React pode atrasar a atualização e então atualizar diversos componentes numa só leva. O React não garante que as mudanças no *state* são aplicadas imediatamente.
```

```

`state` é a referência ao *state* do componente no momento que a mudança está sendo aplicada. Ele não deve ser mutado diretamente. As mudanças devem ser representadas construindo um novo objeto baseado no *input* de *state* e *props*. Por exemplo, suponha que queiramos incrementar um valor no *state* em `props.step`:

```javascript
this.setState((state, props) => {
  return {counter: state.counter + props.step};
});
```

Tanto `state` quanto `props` que foram recebidas pela função `updater` tem a garantia de estarem atualizados. A saída do *updater* é superficialmente mesclada com o *state*.

O segundo parâmetro de `setState()` é uma função de *callback* opcional que será executada assim que `setState` for completada e o componente rerenderizado. Geralmente, recomendamos utilizar `componentDidUpdate()` para implementar esta lógica.

Você também pode passar um objeto como primeiro argumento de `setState()` ao invés de uma função:

```javascript
setState(stateChange[, callback])
```

Isto realiza uma mescla superficial de `stateChange` dentro no novo *state*. Por exemplo: para ajustar a quantidade de items no carrinho de compras:

```javascript
this.setState({quantity: 2})
```

> Esta forma de `setState()` também é assíncrona, e múltiplas chamadas durante o mesmo ciclo podem ser agrupadas numa só. Por exemplo, se você tentar incrementar a quantidade de itens mais que uma vez no mesmo ciclo, isto resultará no equivalente a:

Reviewer 2: ```suggestion
Esta forma de `setState()` também é assíncrona e múltiplas chamadas durante o mesmo ciclo podem ser agrupadas numa só. Por exemplo, se você tentar incrementar a quantidade de itens mais que uma vez no mesmo ciclo, isto resultará no equivalente a:
```

)
```

> Chamadas subsequentes irão sobrescrever valores de chamadas anteriores no mesmo ciclo, então a quantidade será incrementada somente uma vez. Se o próximo estado depende do estado atual, recomendamos utilizar a função *updater* como alternativa:

Reviewer 2: ```suggestion
Chamadas subsequentes irão sobrescrever valores de chamadas anteriores no mesmo ciclo. Com isso, a quantidade será incrementada somente uma vez. Se o próximo estado depende do estado atual, recomendamos utilizar a função *updater* como alternativa:
```

};
```

Se `props.color` não for informado, o seu valor será definido como `'blue'`:

```js
  render() {
    return <CustomButton /> ; // props.color será definido como 'blue'
  }
```

Se `props.color` for igual a `null`, continuará como `null`:

```js
  render() {
    return <CustomButton color={null} /> ; // props.color continuará null
  }
```

* * *

### `displayName` {#displayname}

> A *string* `displayName` é utilizada em mensagens de depuração. Normalmente, você não precisa defini-la explicitamente, pois é isto é inferido do nome da função ou classe que definem o componente. Você pode querer defini-la explicitamente se quiser exibir um nome diferente para propósitos de depuração ou quando você cria um *higher-order component*. Veja [Envolva o Display Name para Facilitar a Depuração](/docs/higher-order-components.html#convention-wrap-the-display-name-for-easy-debugging) para mais detalhes.

Reviewer 2: ```suggestion
A *string* `displayName` é utilizada em mensagens de depuração. Normalmente, você não precisa defini-la explicitamente. Pois isto é inferido do nome da função ou classe que definem o componente. Você pode querer defini-la explicitamente se quiser exibir um nome diferente para propósitos de depuração ou quando você cria um *higher-order component*. Veja [Envolva o Display Name para Facilitar a Depuração](/docs/higher-order-components.html#convention-wrap-the-display-name-for-easy-debugging) para mais detalhes.
```

* * *

## Instance Properties {#instance-properties-1}

### `props` {#props}

`this.props` contém as *props* que foram definidas por quem chamou este componente. Veja [Componentes e Props](/docs/components-and-props.html) para uma introdução às *props*.

> Em particular, `this.props.children` é uma *prop* especial, tipicamente definida pelas *tags* filhas na expressão JSX, ao invés de na própria *tag*.

Reviewer 2: ```suggestion
Em particular, `this.props.children` é uma *prop* especial, normalmente definida pelas *tags* filhas na expressão JSX, ao invés de na própria *tag*.
```

### `state` {#state}

> O *state* contém dados específicos a este componente que podem mudar com o tempo. O *state* é definido pelo usuário, e deve ser um objeto JavaScript.

Reviewer 2: ```suggestion
O *state* contém dados específicos a este componente que podem mudar com o tempo. O *state* é definido pelo usuário e deve ser um objeto JavaScript.
```
